File: optimizer/perturbation.py
import time
import random
import data_api.state            as state_api
import optimizer.rasp_slots      as rasp_slots
import optimizer.tax_tool        as tax_tool
import optimizer.grade_tool      as grade_tool
import data_api.state            as state_api
import optimizer.local_search    as local_search
from tqdm import tqdm


def random_timetable(state):
    print("Generating random timetable.")
    state_api.clear_mutable(state)
    set_random_slots(state, state.rasps)


def semi_random_timetable(state, percent=0.05):
    first_iteration = True if not state.timetable else False
    if first_iteration:
        print("Generating random timetable.")
        set_random_slots(state, state.rasps)
    else:
        print("Generating semi random timetable.")
        set_semi_random_slots(state, percent)


def set_random_slots(state, rasps):
    for rasp in tqdm(rasps):
        pool = rasp_slots.get_rasp_slots(state, rasp)
        slot = random.choice(tuple(pool))
        tax_tool.tax_new_slot(state, rasp, slot)


def set_semi_random_slots(state, percent):
    # Schedule only the most problematic rasps
    problematic_rasps = grade_tool.most_problematic_rasps(state, percent)
    state_api.clear_mutable(state)
    set_random_slots(state, problematic_rasps)

    # Reach local optimum
    start_time = time.time()
    local_search.run(state, 5, start_time)

    # Randomly schedule other rasps
    other_rasps = set(state.rasps) - set(problematic_rasps)
    set_random_slots(state, other_rasps)

File: cpp_optimizer/optimizer/perturbation.cpp
#include "perturbation.h"
#include "../types/state.h"
#include "../optimizer/local_search.h"
#include "../taxing/tax_tool.h"
#include "../taxing/rasp_problems.h"
#include "../rrule_logic/cons_api.h"
#include "../rrule_logic/rasp_slots.h"
#include "../common/common.h"
#include <algorithm>

void set_random_slots(State& state, std::vector<Rasp>& rasps) {
    for(Rasp& rasp : rasps) {
        Slot rnd_slot = get_random_slot(state, rasp);
        tax_new_slot(state, rasp, rnd_slot);
        state.rasps_in_timetable.push_back(rasp);
    }
}

void set_semi_random_slots(State& state, float percent, time_point start_time) {
    std::vector<Rasp> problematic_rasps = most_problematic_rasps(state, percent);
    clear_mutable(state);
    set_random_slots(state, problematic_rasps);

    double SECONDS = elapsed_secs(start_time, time_now()) + 5;
    local_search(state, 1, SECONDS, start_time);

    std::vector<Rasp> other_rasps;
    for(Rasp& rasp : state.rasps) {
        if(std::find(problematic_rasps.begin(), problematic_rasps.end(), rasp) != problematic_rasps.end()) {
            continue;
        }
        other_rasps.push_back(rasp);
    }
    set_random_slots(state, other_rasps);
}

void random_timetable(State& state, time_point start_time) {
    std::cout<<elapsed_secs(start_time, time_now())<<" Generating random timetable.\n";
    clear_mutable(state);
    set_random_slots(state, state.rasps);
}

void semi_random_timetable(State& state, float percent, time_point start_time) {
    bool first_iteration = (state.timetable.empty()) ? true : false;
    if(first_iteration) {
        random_timetable(state, start_time);
    }
    else {
        std::cout<<elapsed_secs(start_time, time_now())<<" Generating semi random timetable.\n";
        set_semi_random_slots(state, percent, start_time);
    }
}
